File: backend/app/services/media_service.py
"""媒体服务：上传、媒体访问鉴权。"""

from __future__ import annotations

import os
import tempfile
from pathlib import Path

from fastapi import UploadFile
from sqlalchemy.orm import Session

from app.config import get_storage_root, settings
from app.models.file_metadata import FileMetadata
from app.models.post_media import PostMedia
from app.models.users import User
from app.schemas.common import AppError, ErrorCode
from app.storage import filekit


def upload_media(db: Session, user: User, file: UploadFile) -> dict:
    """通用上传：校验 -> 同步落盘 -> 写 PostMedia(未绑定) + FileMetadata -> 返回。

    后台任务由路由触发。
    采用分块流式落盘，避免大文件全量读入内存。
    """
    if not file.filename:
        raise AppError(ErrorCode.VALIDATION_ERROR, "未提供文件", 400)

    max_any = max(settings.MEDIA_IMAGE_MAX_MB, settings.MEDIA_VIDEO_MAX_MB) * 1024 * 1024
    media_dir = filekit.get_media_dir()

    # 先流式落盘到临时文件（带大小上限校验）
    tmp_fd, tmp_name = tempfile.mkstemp(dir=media_dir, suffix=".tmp")
    tmp_path = Path(tmp_name)
    os.close(tmp_fd)
    try:
        total = filekit.stream_to_file(file.file, tmp_path, max_any)
    except AppError:
        try:
            tmp_path.unlink()
        except OSError:
            pass
        raise
    except Exception:
        try:
            tmp_path.unlink()
        except OSError:
            pass
        raise AppError(ErrorCode.VALIDATION_ERROR, "读取上传文件失败", 400) from None

    # 从临时文件检测真实格式
    try:
        kind = filekit.detect_kind_from_file(tmp_path)
        filekit.check_size(kind, total)
        if kind == "image":
            fmt = filekit.validate_image_file(tmp_path)
            if fmt == "jpg":
                fmt = "jpeg"
            mime = f"image/{fmt}"
        else:
            fmt = filekit.validate_video_file(tmp_path)
            mime = f"video/{fmt}"
    except AppError:
        try:
            tmp_path.unlink()
        except OSError:
            pass
        raise

    # 重命名为最终文件名
    filename = filekit.generate_filename(fmt)
    abs_path = media_dir / filename
    tmp_path.rename(abs_path)
    storage_root = get_storage_root()
    rel_path = str(abs_path.relative_to(storage_root))
    media = PostMedia(
        post_id=None,
        owner_id=user.id,
        file_path=rel_path,
        thumb_path=None,
        large_path=None,
        filename=filename,
        size=total,
        mime=mime,
        format=fmt,
        kind=kind,
        sort_order=0,
    )
    db.add(media)
    db.flush()
    db.add(
        FileMetadata(
            storage_path=rel_path,
            original_name=file.filename or filename,
            filename=filename,
            size=total,
            mime=mime,
            format=fmt,
            kind=kind,
            owner_id=user.id,
        )
    )
    db.commit()
    db.refresh(media)
    return {
        "media_id": media.id,
        "kind": media.kind,
        "format": media.format,
        "size": media.size,
        "status": "pending",
        "created_at": media.created_at,
    }


def get_media_file(
    db: Session, viewer_id: int | None, post_id: int, media_id: int, spec: str
) -> tuple[Path, str]:
    """鉴权 + 防穿越，返回 (abs_path, content_type)。失败抛 AppError。"""
    if spec not in ("thumb", "large", "original"):
        raise AppError(ErrorCode.VALIDATION_ERROR, "规格参数无效", 400)
    from app.models.posts import Post
    from app.utils.visibility import can_view_post

    post = db.query(Post).filter(Post.id == post_id, Post.deleted_at.is_(None)).first()
    if not post:
        raise AppError(ErrorCode.NOT_FOUND, "动态不存在", 404)
    if not can_view_post(db, viewer_id, post):
        raise AppError(ErrorCode.FORBIDDEN, "无权查看此媒体", 403)
    media = (
        db.query(PostMedia).filter(PostMedia.id == media_id, PostMedia.post_id == post_id).first()
    )
    if not media:
        raise AppError(ErrorCode.NOT_FOUND, "媒体不存在", 404)
    if spec == "original":
        rel: str | None = media.file_path
        mime = media.mime
    elif spec == "thumb":
        rel = media.thumb_path
        mime = "image/webp"
    else:
        rel = media.large_path
        mime = "image/webp"
    if not rel:
        raise AppError(ErrorCode.MEDIA_NOT_READY, "媒体尚未就绪", 404)
    abs_path = filekit.resolve_within_storage(rel)
    if not abs_path:
        raise AppError(ErrorCode.NOT_FOUND, "文件不存在", 404)
    return abs_path, mime

File: backend/internal/service/media_service.go
package service

import (
	"context"
	"errors"
	"io"
	"os"
	"path/filepath"

	"backend/internal/config"
	"backend/internal/model"
	"backend/internal/repository"
	"backend/internal/util"
)

type MediaService struct {
	cfg   *config.Config
	repos *repository.Repositories
}

func NewMediaService(cfg *config.Config, repos *repository.Repositories) *MediaService {
	return &MediaService{cfg: cfg, repos: repos}
}

func (s *MediaService) UploadMedia(ctx context.Context, user *model.User, originalName string, r io.Reader) (*model.MediaUploadOut, error) {
	if originalName == "" {
		return nil, model.NewAppError(model.ErrValidationError, "未提供文件", 400)
	}

	maxAnyMB := s.cfg.MediaImageMaxMB
	if s.cfg.MediaVideoMaxMB > maxAnyMB {
		maxAnyMB = s.cfg.MediaVideoMaxMB
	}
	maxAnyBytes := int64(maxAnyMB) * 1024 * 1024

	mediaDir, err := util.GetMediaDir(s.cfg.StorageRoot)
	if err != nil {
		return nil, model.NewAppError(model.ErrInternal, "创建媒体目录失败", 500)
	}

	// Stream to temp file
	tmpFile, err := os.CreateTemp(mediaDir, "upload_*.tmp")
	if err != nil {
		return nil, model.NewAppError(model.ErrInternal, "创建临时文件失败", 500)
	}
	tmpPath := tmpFile.Name()
	_ = tmpFile.Close()

	total, err := util.StreamToFile(r, tmpPath, maxAnyBytes, s.cfg.StorageRoot)
	if err != nil {
		_ = os.Remove(tmpPath)
		if errors.Is(err, errors.New("file too large")) {
			return nil, model.NewAppError(model.ErrFileTooLarge, "文件过大", 413)
		}
		return nil, model.NewAppError(model.ErrValidationError, "读取上传文件失败", 400)
	}

	// Read header to detect kind
	f, err := os.Open(tmpPath)
	if err != nil {
		_ = os.Remove(tmpPath)
		return nil, model.NewAppError(model.ErrInternal, "打开临时文件失败", 500)
	}
	header := make([]byte, 261)
	n, _ := f.Read(header)
	_ = f.Close()

	kind, ext, mime, err := util.DetectKind(header[:n])
	if err != nil {
		_ = os.Remove(tmpPath)
		return nil, model.NewAppError(model.ErrUnsupportedMedia, "不支持的文件类型", 400)
	}

	// Check kind size
	if kind == "image" {
		if total > int64(s.cfg.MediaImageMaxMB)*1024*1024 {
			_ = os.Remove(tmpPath)
			return nil, model.NewAppError(model.ErrFileTooLarge, "文件过大", 413)
		}
	} else if kind == "video" {
		if total > int64(s.cfg.MediaVideoMaxMB)*1024*1024 {
			_ = os.Remove(tmpPath)
			return nil, model.NewAppError(model.ErrFileTooLarge, "文件过大", 413)
		}
	}

	finalFilename := util.GenerateFilename(ext)
	finalAbsPath := filepath.Join(mediaDir, finalFilename)
	if err := os.Rename(tmpPath, finalAbsPath); err != nil {
		_ = os.Remove(tmpPath)
		return nil, model.NewAppError(model.ErrInternal, "保存媒体文件失败", 500)
	}

	relPath, err := filepath.Rel(s.cfg.StorageRoot, finalAbsPath)
	if err != nil {
		relPath = filepath.Join("media", filepath.Base(mediaDir), finalFilename)
	}

	media := &model.PostMedia{
		PostID:    nil,
		OwnerID:   &user.ID,
		FilePath:  relPath,
		Filename:  finalFilename,
		Size:      int(total),
		Mime:      mime,
		Format:    ext,
		Kind:      kind,
		SortOrder: 0,
	}

	if err := s.repos.Media.CreateMedia(ctx, media); err != nil {
		return nil, model.NewAppError(model.ErrInternal, "记录媒体失败", 500)
	}

	_ = s.repos.Media.CreateFileMetadata(ctx, &model.FileMetadata{
		StoragePath:  relPath,
		OriginalName: originalName,
		Filename:     finalFilename,
		Size:         int(total),
		Mime:         mime,
		Format:       ext,
		Kind:         kind,
		OwnerID:      &user.ID,
	})

	return &model.MediaUploadOut{
		MediaID:   media.ID,
		Kind:      media.Kind,
		Format:    media.Format,
		Size:      media.Size,
		Status:    "pending",
		CreatedAt: media.CreatedAt,
	}, nil
}

func (s *MediaService) GetMediaFile(ctx context.Context, viewerID *int, postID, mediaID int, spec string) (string, string, error) {
	if spec != "thumb" && spec != "large" && spec != "original" {
		return "", "", model.NewAppError(model.ErrValidationError, "规格参数无效", 400)
	}

	post, err := s.repos.Post.GetByID(ctx, postID)
	if err != nil {
		return "", "", err
	}
	if post == nil {
		return "", "", model.NewAppError(model.ErrNotFound, "动态不存在", 404)
	}

	canView := false
	if post.Visibility == "public" {
		canView = true
	} else if viewerID != nil {
		if *viewerID == post.UserID {
			canView = true
		} else {
			areFriends, _ := s.repos.Friend.AreFriends(ctx, *viewerID, post.UserID)
			canView = areFriends
		}
	}
	if !canView {
		return "", "", model.NewAppError(model.ErrForbidden, "无权查看此媒体", 403)
	}

	media, err := s.repos.Media.GetPostMedia(ctx, postID, mediaID)
	if err != nil {
		return "", "", err
	}
	if media == nil {
		return "", "", model.NewAppError(model.ErrNotFound, "媒体不存在", 404)
	}

	var relPath *string
	mime := media.Mime

	switch spec {
	case "original":
		relPath = &media.FilePath
	case "thumb":
		relPath = media.ThumbPath
		mime = "image/webp"
	case "large":
		relPath = media.LargePath
		mime = "image/webp"
	}

	if relPath == nil || *relPath == "" {
		return "", "", model.NewAppError(model.ErrMediaNotReady, "媒体尚未就绪", 404)
	}

	absPath, ok := util.ResolveWithinStorage(s.cfg.StorageRoot, *relPath)
	if !ok {
		return "", "", model.NewAppError(model.ErrNotFound, "文件不存在", 404)
	}

	return absPath, mime, nil
}
